### DISCUSSION TITLE
Comparison with OXC

### DISCUSSION THREAD

**Discussion by notramo** [biomejs/biome#1281] [Q&A] [answered]
I can't really tell which one to choose over the other. Could you put a comparison of the linter and formatter on the website?

**Comment by Boshen** [comment]
Oxc doesn't have a formatter yet.

Give both a try and see which one fits your requirements. 

From my point of view, Biome excels in every aspect at the moment.

**Reply by notramo** [reply]
Then what are the reasons OXC is developed? Will it support functionality that Biome doesn't?

**Reply by mcmxcdev** [reply]
I asked pretty much the same question in https://github.com/oxc-project/oxc/discussions/1709, so it seems to be confusing for devs

**Reply by Boshen** [reply]
Two years have passed since the previous reply.

Our formatter is now ready https://oxc.rs/docs/guide/usage/formatter.html

And I believe Oxc has excelled in most areas today.

**Reply by mrhyde** [reply]
Just tried it yesterday and it worked perfectly! Exactly what I expected from a formatter.
Great job! 🔥

**Comment by Conaclos** [comment, marked as answer]
Biome and OXC are relatively close because OXC is inspired by Rome Tools (now Biome). However, they have some differences. Please take my description with a pinch of salt, because I am part of the Biome team.

## Goals

OXC aims to provide a suite of tools for JavaScript, TypeScript and JSX. It already provides a linter and is working on a formatter and a transpiler.

Biome aims to provide a suite of tools for [all major web languages](https://biomejs.dev/internals/language-support/) with first-class editor support. We currently support JavaScript, TypeScript, JSX and JSON. We are working on CSS support. Biome already provides a linter and a Prettier-compatible formatter. We aim to support a transpiler in the long term.

## Formatter

Biome provides a production-ready formatter that scores 97% compatibility with Prettier. It supports JavaScript, TypeScript, JSX, and JSON. Most of the options of Prettier are available.

If you want a fast Prettier replacement, Biome is the only choice here.

## Linter

OXC attempts to provide a near drop-in replacement for ESLint and some ESLint plugins. It implements ESLInt rules as closely as possible (supporting rules' options).

Biome also provides most of the ESLint rules and some ESLint rules from plugins. However, it's not intended to be a drop-in replacement. This means that Biome has changed some rule names, doesn't support most of the rules' options, and sometimes implements slightly different behavior.

So Biome and OXC are pretty close here. OXC has the advantage of familiarity because it closely mimics ESLint. Biome is trying to provide a new linter experience.

## Performance

Biome and OXC are both performant for most of the projects. For very large projects, I think that OXC is faster than Biome because it is designed from the ground up for performance. In addition, OXC has made different technology choices that also allow for performance gains (see next section).

## Editor supports

Biome and OXC both offer LSP support. However, Biome offers more options and is able to format and lint malformed code. This is one of Biome's main goals. This advantage is due to Biome's technological choices. Biome uses a CST (Concrete Syntax Tree) and an error-resilient parser, whereas OXC uses an AST (Abstract Syntax Tree). The choice of an AST instead of a CST is one of the raisons d'être of OXC. If I remember correctly, @Boshen preferred an AST because of its simplicity.

**Reply by Boshen** [reply]
Thank you for the thorough explanation 👍

**Reply by valerii15298** [reply]
I do not fully understand why not just merge two projects...
The overlap seems quite big.
Merging would bring quite an amount of users and I imagine can allow incorporate best features of both in one package.

If you guys merge I will donate $100 to both each of you 😆 (will simplify my choice to use one tool since now I am using eslint because do not wanna chose oxc or biome and dig into differences, details)

But I feel like it is not as simple to merge...

**Reply by mrhyde** [reply]
I have to strongly disagree with the claim that Biome is "able to format and lint malformed code." This is simply false and frankly misleading. In reality Biome not only fails to handle malformed code gracefully—it often crashes outright when encountering syntax errors or incomplete code structures. I've been using it for past 2 years and honestly it's been a huge letdown. Whoever might reed this in 2026 - look elsewhere, this is not the way.

**Reply by dyc3** [reply]
@mrhyde feel free to send bug reports

**Comment by eightHundreds** [comment]
i need eslint-plugin-biome-oxlint or oxlint-plugin-biome
